### PR TITLE
[FEATURE] Output changelog url in OutputFormatterInterface

### PR DESCRIPTION
**Pull request by sabbelasichon**
Resolves: #6072

### CODE REVIEW

**Comment by sabbelasichon** [comment]
I am cheating here. I don´t know why or how i should use the phpstan ClassReflection here. Any ideas?

File: packages/ChangesReporting/ValueObject/RectorWithFileAndLineChange.php [hunk 1]
```diff
@@ -41,6 +42,33 @@ public function getRectorClass(): string
         return get_class($this->rector);
     }
 
+    public function getRectorClassWithChangelogUrl(): string
+    {
+        $rectorClass = get_class($this->rector);
+
+        /* @phpstan-ignore-next-line */
```
Comment (TomasVotruba) [inline packages/ChangesReporting/ValueObject/RectorWithFileAndLineChange.php:49]: We put ignores to `phsptan.neon`, to keep track on them and avoid code polution with tool ignores. Could you move it there?

**Comment by TomasVotruba** [comment]
Thanks for the PR! 

I have troubles understanding, because "changelog" reminds me `CHANGELOG.md`, but I guess this is probably just showing `@link` in the output after list of affeted rules?

<br>

Could you provide some real life example how you use it and what happens? Gif is worth thousand words :)


```bash
vendor/bin/rector ...
↓
output (screenshot)
```

**Comment by sabbelasichon** [comment]
Screenshot 2021-04-10 at 12 19 29

**Comment by sabbelasichon** [comment]
Changelog has nothing to do in our case with the Changelog.md file. We are creating for every change in TYPO3 a really nice changelog file in the rst format transformed to html. Example: https://docs.typo3.org/c/typo3/cms-core/master/en-us/Changelog/9.4/Deprecation-85793-SeveralConstantsFromSystemEnvironmentBuilder.html. In TYPO3 Rector i add for every rector a nice @link block to reference to this url. This is really nice for the consumer to see why a certain rules is applied.

**Comment by sabbelasichon** [comment]
Screenshot 2021-04-10 at 12 30 32

That´s the json output

**Comment by TomasVotruba** [comment]
I see. So this way we can learn more detailed desription of the change in particular framework?

We actually do something like this in implicit way:

https://github.com/rectorphp/rector/blob/66a39a1ac903e4e090b7cd55914437669df3a192/rules/Php55/Rector/Class_/ClassConstantToSelfClassRector.php#L15

<br>

We could enable this by default, as there is no reason to hide this information.

<br>

One more thing. Currently the JSON output is working with classes only in "applied_rectors". That way it can be used in [Rector demo page](https://getrector.org/demo) to link exact rule: 

image

To keep the original value and add a new one, there should be a new value:

```json
{
    "applied_rules": ["SomeRule"],
    "applied_rules_with_changelog": [
        "SomeRule": "https://project/change/link.md"
    ]
}
```

**Comment by sabbelasichon** [comment]
@TomasVotruba I am not 100% sure about the tests. I have introduced some tests to verify the behaviour. Am i right, that no tests exists before to ensure the desired behaviour? Do i miss something?

File: packages/ChangesReporting/ValueObject/RectorWithFileAndLineChange.php [hunk 2]
```diff
@@ -41,6 +42,44 @@ public function getRectorClass(): string
...
+
+        if (! array_key_exists('url', $matches)) {
+            return null;
+        }
+
+        if (! filter_var($matches['url'], FILTER_VALIDATE_URL)) {
+            return null;
+        }
+
+        return trim((string) $matches['url']);
```
Comment (TomasVotruba) [inline packages/ChangesReporting/ValueObject/RectorWithFileAndLineChange.php:80]: This seems quite a complext logic for simple value object.
Could you extract this to own service and cover it with simple test?

```php
$extractedValue = $this->extractAnnotationFromClass('SomeRector::class', 'link');
$this->assertSame('...', $extractedValue);
```

**Comment by TomasVotruba** [comment]
These are the first test to covert this part, so I think it's ok :+1: 

It's getting into good shape :+1: 

I wonder about last think - the annotation name, @see, @link and @source seems a bit unclear. Now that you explained the purpose, maybe the name `@changelog` would be better name. In most of Symfony/Nette/Laravel/CakePHP rules it would actually link a headline in changelog where the change is described. What do you think?

This feature would be a big addition to understand what rules do and why :+1:

**Comment by sabbelasichon** [comment]
Is @changelog an de-facto official docblock? But, maybe we should not care about it and invent our own.

File: packages/ChangesReporting/ValueObject/RectorWithFileAndLineChange.php (hunk 2 above)
Comment (sabbelasichon) [reply packages/ChangesReporting/ValueObject/RectorWithFileAndLineChange.php:80]: Where should this service be located? Is there not already something similar we could use?

File: packages/ChangesReporting/ValueObject/RectorWithFileAndLineChange.php (hunk 1 above)
Comment (TomasVotruba) [reply packages/ChangesReporting/ValueObject/RectorWithFileAndLineChange.php]: Fixed by: https://github.com/rectorphp/rector/pull/6078

**Comment by TomasVotruba** [comment]
> But, maybe we should not care about it and invent our own.

:+1: I think it's good name for what is does

File: packages/ChangesReporting/ValueObject/RectorWithFileAndLineChange.php (hunk 2 above)
Comment (TomasVotruba) [reply packages/ChangesReporting/ValueObject/RectorWithFileAndLineChange.php:80]: I'll trust your choice :+1:

**Comment by sabbelasichon** [comment]
If i run the phpstan analysis locally everything is fine. Weird.

**Comment by TomasVotruba** [comment]
That's because I've just added this rule into `main` branch. Just ignore it in `phpstan.neon`, I'll handle it later.

**Comment by TomasVotruba** [comment]
Thank you :+1: Let's :ship: it :)

**Pull request merged**